### PR TITLE
Support create_before_destroy for ASG

### PR DESCRIPTION
**Pull request by alexjurkiewicz**
(You probably want to flatten this PR on merge if you accept it -- i created it in github's web interface)

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -8,6 +8,11 @@ variable "create_asg" {
   default     = true
 }
 
+variable "create_before_destroy" {
```
Comment (alexjurkiewicz) [inline variables.tf]: I put this near the top of the options list, I can move it if you like. I also set the default to be the same as current.

I personally think the default should change to `true` :)

**Comment by antonbabenko** [comment]
Why do you need to have `create_before_destroy = true` on ASG? Isn't it enough to have it on launch configuration resource?

**Comment by alexjurkiewicz** [comment]
I'm thinking about cases where I make a change to an ASG property that
requires a rebuild.

**Comment by antonbabenko** [comment]
ASG allows being changed in-place, while launch configurations can't be changed and only new have to be created. This is the reason for having `create_before_destroy = true` on launch configurations.

**Comment by alexjurkiewicz** [comment]
Here's a workflow that I believe `create_before_destroy = true` on ASG would help:
I have ASG with scheduled actions defined on it:
```hcl
resource "aws_autoscaling_schedule" "router-scaledown" {
  # [...]
  autoscaling_group_name = "${module.autoscaling.this_autoscaling_group_name}"
}
```

The current behaviour when I update the LC (eg to update user data) is as follows:
1. Delete LC and all autoscaling schedules; create new LC
2. Create new ASG
3. Wait for new ASG to meet `min_elb_capacity`
4. Create new autoscaling schedules; destroy old ASG
5. Finish successfully

If the new Terraform apply fails (eg the new ASG never becomes healthy), I will be left with an old ASG that has no scheduled actions, which is much harder to fix. Adding `create_before_destroy = true` to the ASG, the behaviour changes to:

1. Create new LC
2. Create new ASG
3. Wait for new ASG to meet `min_elb_capacity`
4. Delete old autoscaling schedules, destroy old asg
5. Finish successfully

**Comment by antonbabenko** [comment]
Lifecycle `create_before_destroy` specified in the module is not going to have an effect on resources created outside of this module, but only inside of the same module. This is why `aws_launch_configuration` and `aws_autoscaling_group` are in the same module and not in separate modules as it was [before](https://github.com/terraform-aws-modules/terraform-aws-autoscaling/blob/b5c431df2bbc8187456c316b59241a1e1f4bea80/main.tf).

We can add support for other types of resources (ASG schedule, for eg), but it will make this module more complicated for basic users, so I am not in favor of this.

I don't see how this module can be changed to support flexibility required for blue-green deployments.

**Comment by alexjurkiewicz** [comment]
hm, maybe it's possible to add `create_before_destroy` to the schedule
resource after making it explicitly depend on the ASG... I'll have to test.

**Comment by antonbabenko** [comment]
`create_before_destroy` does not work between modules either.

**Comment by alexjurkiewicz** [comment]
😭

**Comment by antonbabenko** [comment]
[v2.6.0](https://github.com/terraform-aws-modules/terraform-aws-autoscaling/releases/v2.6.0) has been released and it contains this fix.

**Pull request closed without merging**